### ISSUE TITLE
Structured Denoising Diffusion Models in Discrete State-Spaces

### ISSUE THREAD

**Issue by shnakazawa** [TeamAidemy/ds-paper-summaries#52]
Austin, Jacob, Daniel D. Johnson, Jonathan Ho, Daniel Tarlow, and Rianne van den Berg. 2021. “Structured Denoising Diffusion Models in Discrete State-Spaces.” arXiv [cs.LG]. arXiv. http://arxiv.org/abs/2107.03006.

Advances in Neural Information Processing Systems 34 (NeurIPS 2021) に採択
https://proceedings.neurips.cc//paper/2021/hash/958c530554f78bcd8e97125b70e6973d-Abstract.html

- 拡散生成モデル、主に研究が進んでいたのは連続確率分布を用いたものだったが、離散確率分布を用いてそこそこ良い性能のものをつくることに成功。
	- Discrete Denoising Diffusion Probabilistic Models で D3PMs
		- Model**s**と複数形になっているのは、用いる確率分布が何種類かあるため。それによって得意とするタスクが変わる。
- 離散確率分布を用いることで、離散データであるテキストの生成ができるようになった。
- 拡散モデルの枠組みにBERTやGPTといった強力な言語モデルをはめ込むことができると議論。

## Abstract

> Denoising diffusion probabilistic models (DDPMs) (Ho et al. 2020) have shown impressive results on image and waveform generation in continuous state spaces. Here, we introduce Discrete Denoising Diffusion Probabilistic Models (D3PMs), diffusion-like generative models for discrete data that generalize the multinomial diffusion model of Hoogeboom et al. 2021, by going beyond corruption processes with uniform transition probabilities. This includes corruption with transition matrices that mimic Gaussian kernels in continuous space, matrices based on nearest neighbors in embedding space, and matrices that introduce absorbing states. The third allows us to draw a connection between diffusion models and autoregressive and mask-based generative models. We show that the choice of transition matrix is an important design decision that leads to improved results in image and text domains. We also introduce a new loss function that combines the variational lower bound with an auxiliary cross entropy loss. For text, this model class achieves strong results on character-level text generation while scaling to large vocabularies on LM1B. On the image dataset CIFAR-10, our models approach the sample quality and exceed the log-likelihood of the continuous-space DDPM model.

(DeepL翻訳)

ノイズ除去拡散確率モデル(DDPM) (Ho et al. 2020)は連続状態空間における画像や波形の生成で印象的な結果を示している。ここでは、一様な遷移確率を持つ破損過程を超えて、Hoogeboomら2021の多項拡散モデルを一般化した離散データの拡散的生成モデルである離散ノイズ拡散確率モデル（D3PMs）を紹介する。これには、連続空間におけるガウスカーネルを模倣した遷移行列、埋め込み空間における最近接に基づく行列、吸収状態を導入した行列を用いたコラプションが含まれる。3つ目は、拡散モデルと自己回帰モデルやマスクベースの生成モデルとの間の関連性を描くことができる。我々は、遷移行列の選択が、画像やテキスト領域での結果の改善につながる重要な設計上の決定であることを示す。また、変分下界と補助的なクロスエントロピー損失を組み合わせた新しい損失関数を導入する。テキスト分野では、このモデルクラスはLM1Bの大語彙に対応しながら、文字レベルのテキスト生成で強い結果を達成した。画像データセットCIFAR-10では、我々のモデルはサンプル品質に近づき、連続空間DDPMモデルの対数尤度を上回る。

## コード

https://github.com/samb-t/unleashing-transformers

## 解決した課題/先行研究との比較

- 拡散生成モデルが画像・音声生成の分野では目覚ましい成果を上げている。
- 多くの拡散モデルはガウシアン拡散過程 (= 連続分布) に立脚している
	- 本論文以前の離散データ生成の取り組み、離散分布を用いた取り組みについては下記の[重要な引用](#重要な引用) 参照。
- 本論文では**新しいタイプの離散拡散過程を定義**することにより、カテゴリカル・離散的なデータの拡散生成モデルを改良。
	- それにより離散的データであるテキストの生成に成功。
- Discrete Denoising Diffusion Probabilistic Models で **D3PMs** と命名

## 技術・手法のポイント

- データにノイズを加える拡散過程にて、データやノイズを離散値・離散確率分布として扱う。
- ノイズが加わる = 文章の一部のトークンが [MASK] に置き換えられる。
- 加えるノイズのパターンも4種類検討 (**Section 3.1**)
	- 一様分布 (Uniform)
	- 吸収状態 (absorbing)
	- 離散ガウシアン (Discretized Gaussian)
	- トークン埋め込み距離 (Token embedding distance) ... nearest neighbors を用いる(NN)
[![Image from Gyazo](https://i.gyazo.com/f669fd7da5167702ab490840961e320f.png)](https://gyazo.com/f669fd7da5167702ab490840961e320f)
- 損失関数には通常の拡散モデルで使われる variational lower bound (VLB: 変分下限) に加え、auxiliary loss を導入
	- それぞれのステップから元画像が復元できるか？の指標
- この拡散モデルの考え方を広げればBERTやGPTを結びつけて説明することができる。
	- 例えば、BERTはノイズを1段階載せた拡散モデルと等価と考えられる。
	- 詳しくは **Section 4** を参照。面白い。

## 評価指標

- 言語タスク
	- text8, LM1Bデータセットを用い、各モデルの比較。(**Tables 1,2**)
		- D3PMsの中では吸収状態 (absorbing state) を使った場合が最も性能が良かった
		- しかし、Transformerの性能には及ばない。
- 画像タスク
	- CIFAR-10データセットを用いて画像生成の枠組でも本論文の手法を評価。(**Table 3**)
	- D3PMsの中では離散ガウシアンを使った場合が最も性能が良かった。
	- スコアベースモデル (Song and Ermon, 2019, 2020)、拡散モデル (Ho et al., 2020) と同程度の性能。
- → **タスクに応じて最適な確率分布の形が異なる**

## 残された課題・議論

- 生成物の出来はそれほど良くない。テキストならTransformerのような自己回帰モデル、画像であれば連続拡散モデルの方が良い出力をする。
	- ステップ数を増やしたり、ノイズの与え方を工夫することで、D3PMs の性能が良くなるかも知れない (as discussed in Nichol and Dhariwal, 2020)

## 重要な引用

- アイデアのもととなった拡散確率モデル
	- Sohl-Dickstein, Jascha, Eric Weiss, Niru Maheswaranathan, and Surya Ganguli. 07--09 Jul 2015. “Deep Unsupervised Learning Using Nonequilibrium Thermodynamics.” In Proceedings of the 32nd International Conference on Machine Learning, edited by Francis Bach and David Blei, 37:2256–65. Proceedings of Machine Learning Research. Lille, France: PMLR.
	- Ho, Jonathan, Ajay Jain, and Pieter Abbeel. 2020. “Denoising Diffusion Probabilistic Models.” arXiv [cs.LG]. arXiv. http://arxiv.org/abs/2006.11239.
- BERT
	- Devlin, Jacob, Ming-Wei Chang, Kenton Lee, and Kristina Toutanova. 2018. “BERT: Pre-Training of Deep Bidirectional Transformers for Language Understanding.” arXiv [cs.CL]. arXiv. http://arxiv.org/abs/1810.04805.
- 拡散モデルに離散確率分布を用いようとした/離散タスクを解決しようとした仕事たち
	- > While most works have considered continuous diffusion models, discrete diffusion-like models were described in [36] and applied to text generation and image segmentation data in [18].
		- [36] Sohl-Dickstein, Jascha, Eric Weiss, Niru Maheswaranathan, and Surya Ganguli. 07--09 Jul 2015. “Deep Unsupervised Learning Using Nonequilibrium Thermodynamics.” In Proceedings of the 32nd International Conference on Machine Learning, edited by Francis Bach and David Blei, 37:2256–65. Proceedings of Machine Learning Research. Lille, France: PMLR.
		- [18] Hoogeboom, Emiel, Didrik Nielsen, Priyank Jaini, Patrick Forré, and Max Welling. 2021. “Argmax Flows and Multinomial Diffusion: Learning Categorical Distributions.” arXiv [stat.ML]. arXiv. http://arxiv.org/abs/2102.05379.
	- > Some works [29, 27] have dealt with discrete data by embedding it in continuous space and leveraging Gaussian diffusion, but have not applied this to text.
		- [29] Niu, Chenhao, Yang Song, Jiaming Song, Shengjia Zhao, Aditya Grover, and Stefano Ermon. 2020. “Permutation Invariant Graph Generation via Score-Based Generative Modeling.” arXiv [cs.LG]. arXiv. http://arxiv.org/abs/2003.00638.
		- [27] Mittal, Gautam, Jesse Engel, Curtis Hawthorne, and Ian Simon. 2021. “Symbolic Music Generation with Diffusion Models.” arXiv [cs.SD]. arXiv. http://arxiv.org/abs/2103.16091.
	- > Seff et al. [35] considered generation of discrete structured objects using a diffusion-like Markov corruption process
		- [35] Seff, Ari, Wenda Zhou, Farhan Damani, Abigail Doyle, and Ryan P. Adams. 2019. “Discrete Object Generation with Reversible Inductive Construction.” arXiv [cs.LG]. arXiv. http://arxiv.org/abs/1907.08268.
- 拡散モデルの最近の改良
	- Nichol, Alex, and Prafulla Dhariwal. 2021. “Improved Denoising Diffusion Probabilistic Models.” arXiv [cs.LG]. arXiv. http://arxiv.org/abs/2102.09672.

## 参考情報

- [ステート・オブ・AIガイド - GANを超え、言語にも進出 拡散モデルによる画像言語生成の進歩がすごい](https://ja.stateofaiguides.com/20210824-diffusion-models/)
